flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/TestFlinkUpsert.java

```diff
@@ -0,0 +1,155 @@
+          Namespace baseNamespace = (Namespace) catalogParams[1];
+          parameters.add(new Object[] {catalogName, baseNamespace, format, isStreaming});
+        }
+      }
+    }
+    return parameters;
+  }
+
+  public TestFlinkUpsert(String catalogName, Namespace baseNamespace, FileFormat format, Boolean isStreamingJob) {
+    super(catalogName, baseNamespace);
+    this.format = format;
+    this.isStreamingJob = isStreamingJob;
+    tableUpsertProps.put(TableProperties.FORMAT_VERSION, "2");
+    tableUpsertProps.put(TableProperties.UPSERT_ENABLED, "true");
+    tableUpsertProps.put(TableProperties.DEFAULT_FILE_FORMAT, format.name());
+  }
+
+  @Override
+  protected TableEnvironment getTableEnv() {
+    if (tEnv == null) {
+      synchronized (this) {
+        EnvironmentSettings.Builder settingsBuilder = EnvironmentSettings
+            .newInstance();
+        if (isStreamingJob) {
+          settingsBuilder.inStreamingMode();
+          StreamExecutionEnvironment env = StreamExecutionEnvironment
+              .getExecutionEnvironment(MiniClusterResource.DISABLE_CLASSLOADER_CHECK_CONFIG);
+          env.enableCheckpointing(400);
+          env.setMaxParallelism(2);
+          env.setParallelism(2);
+          tEnv = StreamTableEnvironment.create(env, settingsBuilder.build());
+        } else {
+          settingsBuilder.inBatchMode();
+          tEnv = TableEnvironment.create(settingsBuilder.build());
+        }
+      }
+    }
+    return tEnv;
+  }
+
+  @Override
+  @Before
+  public void before() {
+    super.before();
+    sql("CREATE DATABASE %s", flinkDatabase);
+    sql("USE CATALOG %s", catalogName);
+    sql("USE %s", DATABASE);
+  }
+
+  @Override
+  @After
+  public void clean() {
+    sql("DROP DATABASE IF EXISTS %s", flinkDatabase);
+    BoundedTableFactory.clearDataSets();
+    super.clean();
+  }
+
+  @Test
+  public void testUpsertAndQuery() {
+    String tableName = "test_upsert_query";
+
+    sql("CREATE TABLE %s(id INT NOT NULL, province STRING NOT NULL, dt DATE, PRIMARY KEY(id,province) NOT ENFORCED) " +
+            "PARTITIONED BY (province) WITH %s",
+        tableName, toWithClause(tableUpsertProps));
+
+    sql("INSERT INTO %s VALUES " +
+        "(1, 'a', TO_DATE('2022-03-01'))," +
+        "(2, 'b', TO_DATE('2022-03-01'))," +
+        "(1, 'b', TO_DATE('2022-03-01'))",
+        tableName);
+
+    sql("INSERT INTO %s VALUES " +
+        "(4, 'a', TO_DATE('2022-03-02'))," +
+        "(5, 'b', TO_DATE('2022-03-02'))," +
+        "(1, 'b', TO_DATE('2022-03-02'))",
+        tableName);
+
+    List<Row> result = sql("SELECT * FROM %s WHERE dt < '2022-03-02'", tableName);
+
+    Assert.assertEquals("result should have 2 rows!", 2, result.size());
+
+    result = sql("SELECT * FROM %s WHERE dt < '2022-03-03'", tableName);
+
+    Assert.assertEquals("result should have 5 rows!", 5, result.size());
+
+    sql("DROP TABLE IF EXISTS %s.%s", flinkDatabase, tableName);
```

Comment:
Nit: Please make sure the table gets dropped, even with failures etc.

Two options:
1) Use `try … finally`
2) Create and drop table in before / after functions (or any of `BeforeEach` etc).

Option 2 is preferable if possible, but it’s not always possible in which case 1 at least makes sure the DROP TABLE statement runs no matter what.